File: src/adapters/lock/redis_lock.ts
import { randomUUID } from 'node:crypto';
import type { LockPort } from '../../app/ports/lock_port';
import type { LoggerPort } from '../../app/ports/logger_port';

const RUNNER_LOCK_KEY = 'lock:runner';

interface RedisLike {
  set(...args: unknown[]): Promise<string | null>;
  get(...args: unknown[]): Promise<string | null>;
  del(...args: unknown[]): Promise<number>;
}

export class RedisLockAdapter implements LockPort {
  private runnerLockToken: string | null = null;

  constructor(
    private readonly redis: RedisLike,
    private readonly logger: LoggerPort
  ) {}

  async acquireRunnerLock(ttlSeconds: number): Promise<boolean> {
    const token = randomUUID();
    const result = await this.redis.set(RUNNER_LOCK_KEY, token, {
      NX: true,
      EX: ttlSeconds
    });

    if (result === 'OK') {
      this.runnerLockToken = token;
      return true;
    }

    return false;
  }

  async releaseRunnerLock(): Promise<void> {
    if (!this.runnerLockToken) {
      return;
    }

    const currentToken = await this.redis.get(RUNNER_LOCK_KEY);
    if (currentToken === this.runnerLockToken) {
      await this.redis.del(RUNNER_LOCK_KEY);
    } else {
      this.logger.warn('Runner lock token mismatch on release');
    }

    this.runnerLockToken = null;
  }

  async markBarProcessed(barCloseTimeIso: string, ttlSeconds: number): Promise<boolean> {
    const key = `idem:signal:${barCloseTimeIso}`;
    const result = await this.redis.set(key, '1', {
      NX: true,
      EX: ttlSeconds
    });

    return result === 'OK';
  }

  async setInflightTx(signature: string, ttlSeconds: number): Promise<void> {
    await this.redis.set(`tx:inflight:${signature}`, '1', {
      EX: ttlSeconds
    });
  }

  async clearInflightTx(signature: string): Promise<void> {
    await this.redis.del(`tx:inflight:${signature}`);
  }
}

File: pybot/adapters/lock/redis_lock.py
from __future__ import annotations

from uuid import uuid4

from redis import Redis

from pybot.app.ports.lock_port import LockPort
from pybot.app.ports.logger_port import LoggerPort

RUNNER_LOCK_KEY_PREFIX = "lock:runner"
INFLIGHT_TX_KEY_PREFIX = "tx:inflight"
RUNNER_LOCK_RELEASE_SCRIPT = """
if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("del", KEYS[1])
end
return 0
"""


class RedisLockAdapter(LockPort):
    def __init__(self, redis: Redis, logger: LoggerPort, lock_namespace: str):
        self.redis = redis
        self.logger = logger
        self.lock_namespace = lock_namespace
        self.runner_lock_token: str | None = None

    def _runner_lock_key(self) -> str:
        return f"{RUNNER_LOCK_KEY_PREFIX}:{self.lock_namespace}"

    def _entry_idem_key(self, bar_close_time_iso: str) -> str:
        return f"idem:entry:{self.lock_namespace}:{bar_close_time_iso}"

    def _inflight_tx_key(self, signature: str) -> str:
        return f"{INFLIGHT_TX_KEY_PREFIX}:{self.lock_namespace}:{signature}"

    def acquire_runner_lock(self, ttl_seconds: int) -> bool:
        token = str(uuid4())
        result = self.redis.set(self._runner_lock_key(), token, nx=True, ex=ttl_seconds)
        if result:
            self.runner_lock_token = token
            return True
        return False

    def release_runner_lock(self) -> None:
        if self.runner_lock_token is None:
            return
        runner_lock_key = self._runner_lock_key()
        runner_lock_token = self.runner_lock_token
        try:
            released = self.redis.eval(
                RUNNER_LOCK_RELEASE_SCRIPT,
                1,
                runner_lock_key,
                runner_lock_token,
            )
            if released != 1:
                self.logger.warn("Runner lock token mismatch on release")
        except Exception as error:
            self.logger.warn(
                "Runner lock release failed",
                {"error": str(error), "lock_key": runner_lock_key},
            )
        self.runner_lock_token = None

    def mark_entry_attempt(self, bar_close_time_iso: str, ttl_seconds: int) -> bool:
        key = self._entry_idem_key(bar_close_time_iso)
        result = self.redis.set(key, "1", nx=True, ex=ttl_seconds)
        return bool(result)

    def has_entry_attempt(self, bar_close_time_iso: str) -> bool:
        key = self._entry_idem_key(bar_close_time_iso)
        return self.redis.get(key) is not None

    def clear_entry_attempt(self, bar_close_time_iso: str) -> None:
        key = self._entry_idem_key(bar_close_time_iso)
        self.redis.delete(key)

    def set_inflight_tx(self, signature: str, ttl_seconds: int) -> None:
        self.redis.set(self._inflight_tx_key(signature), "1", ex=ttl_seconds)

    def has_inflight_tx(self, signature: str) -> bool:
        return self.redis.get(self._inflight_tx_key(signature)) is not None

    def clear_inflight_tx(self, signature: str) -> None:
        self.redis.delete(self._inflight_tx_key(signature))
